<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Шаг-1.-Обзор-данных" data-toc-modified-id="Шаг-1.-Обзор-данных-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Шаг 1. Обзор данных</a></span></li><li><span><a href="#Шаг-2.1-Заполнение-пропусков" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Шаг 2.1 Заполнение пропусков</a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления.-0.3"><span class="toc-item-num">0.3&nbsp;&nbsp;</span>Шаг 2.2 Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных.-0.4"><span class="toc-item-num">0.4&nbsp;&nbsp;</span>Шаг 2.3. Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов.-0.5"><span class="toc-item-num">0.5&nbsp;&nbsp;</span>Шаг 2.4. Удаление дубликатов.</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-0.6"><span class="toc-item-num">0.6&nbsp;&nbsp;</span>Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода.-0.7"><span class="toc-item-num">0.7&nbsp;&nbsp;</span>Шаг 2.6. Категоризация дохода.</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита." data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита.-0.8"><span class="toc-item-num">0.8&nbsp;&nbsp;</span>Шаг 2.7. Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-0.9"><span class="toc-item-num">0.9&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вопрос-1:" data-toc-modified-id="Вопрос-1:-0.9.0.1"><span class="toc-item-num">0.9.0.1&nbsp;&nbsp;</span>Вопрос 1:</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Общий вывод:</a></span></li><li><span><a href="#Общий-вывод-по-проекту" data-toc-modified-id="Общий-вывод-по-проекту-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Общий вывод по проекту</a></span></li><li><span><a href="#Вторая-итерация-ревью" data-toc-modified-id="Вторая-итерация-ревью-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вторая итерация ревью</a></span></li></ul></div>

# Исследование надежности заемщиков.

<b>Описание проекта</b>

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<b>Описание данных</b>

- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

<b>Цель: выявить следующие зависимости</b>
- Есть ли зависимость между количеством детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?

### Шаг 1. Обзор данных

In [1]:
import pandas as pd

In [2]:
clients = pd.read_csv('data.csv')
clients.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [3]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**отсутствуют значения в колонках days_employed и total_income**  #причины для пропусков могут быть следующие:
#эти люди никогда не работали, поэтому и не получали зп
#либо они просто не указали данные о работе, в связи с этим и данные о зп отсутствуют. т.е. пропуски могут быть как случайными, так и нет

In [4]:
#clients.loc[(clients['days_employed']=='NaN')&(clients['total_income']=='NaN')]
clients[clients['days_employed'].isna()].head(100)

children  days_employed  dob_years education  education_id  \
12          0            NaN         65   среднее             1   
26          0            NaN         41   среднее             1   
29          0            NaN         63   среднее             1   
41          0            NaN         50   среднее             1   
55          0            NaN         54   среднее             1   
..        ...            ...        ...       ...           ...   
959         0            NaN         25   среднее             1   
973         0            NaN         49   среднее             1   
974         0            NaN         60   среднее             1   
978         3            NaN         28   среднее             1   
980         0            NaN         53    высшее             0   

             family_status  family_status_id gender  income_type  debt  \
12        гражданский брак                 1      M    пенсионер     0   
26         женат / замужем                 0      M  госслужащий     0   
29   Не женат / не замужем                 4      F    пенсионер     0   
41         женат / замужем                 0      F  госслужащий     0   
55        гражданский брак                 1      F    пенсионер     1   
..                     ...               ...    ...          ...   ...   
959       гражданский брак                 1      M    сотрудник     0   
973        женат / замужем                 0      F    сотрудник     0   
974        женат / замужем                 0      M    сотрудник     0   
978       гражданский брак                 1      F    сотрудник     0   
980        женат / замужем                 0      F    сотрудник     0   

     total_income                           purpose  
12            NaN                   сыграть свадьбу  
26            NaN                       образование  
29            NaN  строительство жилой недвижимости  
41            NaN  сделка с подержанным автомобилем  
55            NaN                   сыграть свадьбу  
..            ...                               ...  
959           NaN                       образование  
973           NaN                 операции с жильем  
974           NaN                высшее образование  
978           NaN                      ремонт жилью  
980           NaN             получение образования  

[100 rows x 12 columns]

тут я пыталась понять, совпадают ли строчки с NaN в этих двух столбцах.
вроде как да)

In [5]:
clients.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

1. количество детей = -1 физически невозможно
2. количество детей = 20 что-то многовато
3. dob_years минимальное значение 0, вероятнее всего просто не указан возраст
4. отрицательные значения колонки days_employed, может быть ошибка в знаке?
5. в этом же столбце максимальное значение 401755 = 1100,7 лет, а среднее 63046 = 172,7 лет нужно искать ошибку. Также логичнее было бы в таблице перевести этот столбец в годах
6. не описано в чем представлена зарплата, в условных единицах например 

### Шаг 2.1 Заполнение пропусков

In [6]:
print("Количество строк с 'days_employed' < 0:", clients[clients['days_employed'] < 0].shape[0])

Количество строк с 'days_employed' < 0: 15906


1. Всего days_employed 19351, а 15906 из них отрицательные, т.е. почти 82%
единственное предположение, это убрать минус, т.е.использовать abs

In [7]:
clients.isna().sum() # нашли пропуски в столбцах days_employed, total_income  2174           
a = clients['days_employed'].isna().sum()/clients['children'].count() #примерно 10% составляют пропущенные значения для столбца days_employed
b = clients['total_income'].isna().sum()/clients['children'].count()  #примерно 10% составляют пропущенные значения для столбца total_income
print('Доля пропущенных значений для столбца days_employed равна {:.0%}'.format(a))
print('Доля пропущенных значений для столбца total_income равна {:.0%}'.format(b))

Доля пропущенных значений для столбца days_employed равна 10%
Доля пропущенных значений для столбца total_income равна 10%


1. для начала я перевести все отрицательные значения в положительные
2. затем я заполню пропуски медианным значением.
С одной стороны использование именно медианного значения позволяет не удалять полностью информацию о людях, а использовать хоть какие-то данные, в этом случае медианна покажет более достоверное значение, так как если брать среднее, то слишком маленькое или слишком большое значение сильно повлияет на него. Но вообще-то это нечестно, так как при отсутствии информации о работе и зарплате шанс выдачи кредита маленький, а если заполнить медианным значением, но шанс наоборот повышается (я бы приравняла к 0).

Еще нужно сгруппировать данные по типу занятости, посчитать медианы и заполнить пропуски ими. (not done)

In [8]:
#делаем колонку days_employed положительной 
clients['days_employed'] = clients['days_employed'].abs()
#для того, чтобы заполнить пропуски, нужно найти медианные значения для этих 2х столбцов
med_days_employed=clients['days_employed'].sort_values(ascending=False).median() #получилось отрицательное...
med_total_income=clients['total_income'].sort_values(ascending=False).median()
#заполним пропуски
clients['days_employed'] = clients['days_employed'].fillna(med_days_employed)
clients['total_income'] = clients['total_income'].fillna(med_total_income)
clients.info() #чекнули, пропуски заполнены

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [9]:
# Код ревьюера
clients['days_employed'].median()

2194.220566878695

In [10]:
#поменяю количество рабочих дней в года в новой колонке. пока непонятно, что делать с тем, кто работал >50-60 лет
clients['years_employed'] = clients['days_employed']/365

clients[clients['years_employed']>50]['years_employed'].count() #примерно столько человек не похоже, что могли работать так долго

3446

### Шаг 2.2 Проверка данных на аномалии и исправления.

In [11]:
print("Всего людей с количеством детей -1 :", clients[clients['children'] == -1].count()[0])

print("Всего людей с количеством детей 20:", clients[clients['children'] == 20].count()[0])

print("Всего людей с возрастом 0:", clients[clients['dob_years'] == 0].count()[0])

Всего людей с количеством детей -1 : 47
Всего людей с количеством детей 20: 76
Всего людей с возрастом 0: 101


1. детей -1, может быть как +1, может что их нет
2. детей 20 скорее ошибка, либо 2, либо 0, но может и правда так, оставлю, вдруг кому-то действительно так повезло :D
3. возраст 0 скорее ошибка заполнения



In [12]:
#меняем children на 0
clients['children'] = clients['children'].replace(-1,0)

In [13]:
#так как мы не знаем, что это за люди с возрастом 0, заменим на среднее, так как нулеых значений всего 101
mean_dob_years = clients['dob_years'].mean()
clients['dob_years'] = clients['dob_years'].replace(0,mean_dob_years)

### Шаг 2.3. Изменение типов данных.

In [14]:
clients['total_income'] = clients['total_income'].astype('int')

в целом нам также не очень важно точное количество рабочих дней, поэтому повторим процедуру и для этого столбца

In [15]:
clients['days_employed'] = clients['days_employed'].astype('int')

### Шаг 2.4. Удаление дубликатов.

Дубликаты могут повлиять на анализ, особенно если в них будут нестандартные значения (слишком большие или маленькие)

In [16]:
#начнем с поиска явных дубликатов
clients.duplicated().sum()
#удаляем явные дубликаты
clients = clients.drop_duplicates().reset_index(drop=True)
clients.duplicated().sum() #явные дубликаты удалены

0

In [17]:
#теперь найдем неявные дубликаты в столбце education
clients['education'] = clients['education'].str.lower() #привели к нижнему регистру, теперь можно искать
clients['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

In [18]:
#проверим остальные столбцы с группами 
display(clients['family_status'].unique())
display(clients['gender'].unique())
display(clients['income_type'].unique())
display(clients['purpose'].value_counts())

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

array(['F', 'M', 'XNA'], dtype=object)

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

свадьба                                   793
на проведение свадьбы                     773
сыграть свадьбу                           769
операции с недвижимостью                  675
покупка коммерческой недвижимости         662
покупка жилья для сдачи                   652
операции с жильем                         652
операции с коммерческой недвижимостью     650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          625
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

1. В столбце family_status только уникальные значения
2. В столбце gender только уникальные значения
3. В столбце income_type только уникальные значения
4. В столбце purpose очень много похожих целей для взятия кредита, заменим на уникальные

In [20]:
def replace_wrong_values(wrong_values, correct_value): # на вход функции подаются список неправильных значений и строка с правильным значением
    for wrong_value in wrong_values: # перебираем неправильные причины
        clients['purpose'] = clients['purpose'].replace(wrong_value, correct_value) # и для каждого неправильного имени вызываем метод replace()

duplicates1 = ['на проведение свадьбы', 'сыграть свадьбу'] # список неправильных имён
purpose1 = 'свадьба' # правильная причина
duplicates2 = ['покупка коммерческой недвижимости', 'операции с жильем', 'покупка жилья для сдачи','операции с коммерческой недвижимостью','покупка жилья','жилье','покупка жилья для семьи','строительство собственной недвижимости','недвижимость','операции со своей недвижимостью','строительство жилой недвижимости','покупка недвижимости','покупка своего жилья','строительство недвижимости','ремонт жилью','покупка жилой недвижимости'] # список неправильных имён
purpose2 = 'операции с недвижимостью' # правильная причина
duplicates3 = ['на покупку своего автомобиля', 'сделка с подержанным автомобилем','свой автомобиль','на покупку подержанного автомобиля','автомобили','на покупку автомобиля','приобретение автомобиля','сделка с автомобилем'] # список неправильных имён
purpose3 = 'автомобиль' # правильная причина
duplicates4 = ['заняться высшим образованием', 'дополнительное образование','высшее образование','получение дополнительного образования','получение образования','профильное образование','получение высшего образования','заняться образованием'] # список неправильных имён
purpose4 = 'образование' # правильная причина
replace_wrong_values(duplicates1, purpose1)
replace_wrong_values(duplicates2, purpose2)
replace_wrong_values(duplicates3, purpose3)
replace_wrong_values(duplicates4, purpose4)

In [21]:
#проверим количество уникальных целей
clients['purpose'].value_counts()

операции с недвижимостью    10814
автомобиль                   4308
образование                  4014
свадьба                      2335
Name: purpose, dtype: int64

Итак, были удалены явные и неявные дубликаты.
Явные дубликаты могут повлиять на конечный анализ результатов,
неявные дубликаты усложняют группировку данных и делают анализ данных сложным и невозможным.
Сами дубликаты могут возникать из-за невнимательности заполнителя, из-за отсутствия правил заполнения, особенно при вольной форме заполнения.

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [22]:
#создадим два словаря 
edu_log = clients[['education','education_id']]
fam_log = clients[['family_status','family_status_id']]
#удалим дубликаты, чтобы посмотреть соответствие значения и его id
edu_log = edu_log.drop_duplicates().reset_index(drop=True)
fam_log = fam_log.drop_duplicates().reset_index(drop=True)
display(edu_log)
display(fam_log)

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

In [23]:
#удалим из исходной талицы два столбца
#я закомментила удаление, потому что в конце для создания pivot table мне нужен столбец с family_status
#clients.drop(columns = ['education','family_status'], axis = 1, inplace=True) #axis1 для столбцов, inplace чтобы изменить исходную таблицу
clients.head(10) #проверили удаление

children  days_employed  dob_years education  education_id  \
0         1           8437       42.0    высшее             0   
1         1           4024       36.0   среднее             1   
2         0           5623       33.0   среднее             1   
3         3           4124       32.0   среднее             1   
4         0         340266       53.0   среднее             1   
5         0            926       27.0    высшее             0   
6         0           2879       43.0    высшее             0   
7         0            152       50.0   среднее             1   
8         2           6929       35.0    высшее             0   
9         0           2188       41.0   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                    purpose  years_employed  
0  операции с недвижимостью       23.116912  
1                автомобиль       11.026860  
2  операции с недвижимостью       15.406637  
3               образование       11.300677  
4                   свадьба      932.235814  
5  операции с недвижимостью        2.537495  
6  операции с недвижимостью        7.888225  
7               образование        0.418574  
8                   свадьба       18.985932  
9  операции с недвижимостью        5.996593

### Шаг 2.6. Категоризация дохода.

In [24]:
#напишем функции для разделения на группы
def income_group(income):
    
    if 0<=income <= 30000:
        return 'E'
    if 30001<=income <= 50000:
        return 'D'
    if 50001<=income <= 200000:
        return 'C'
    if 200001<=income <= 1000000:
        return 'B'
    return 'A' #предполагая, что все значения в столбце положительные

In [25]:
#создадим новую колонку применив метод apply для вызова функции income_group
clients['total_income_category'] = clients['total_income'].apply(income_group)
clients.head(10) #появились категории

children  days_employed  dob_years education  education_id  \
0         1           8437       42.0    высшее             0   
1         1           4024       36.0   среднее             1   
2         0           5623       33.0   среднее             1   
3         3           4124       32.0   среднее             1   
4         0         340266       53.0   среднее             1   
5         0            926       27.0    высшее             0   
6         0           2879       43.0    высшее             0   
7         0            152       50.0   среднее             1   
8         2           6929       35.0    высшее             0   
9         0           2188       41.0   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                    purpose  years_employed total_income_category  
0  операции с недвижимостью       23.116912                     B  
1                автомобиль       11.026860                     C  
2  операции с недвижимостью       15.406637                     C  
3               образование       11.300677                     B  
4                   свадьба      932.235814                     C  
5  операции с недвижимостью        2.537495                     B  
6  операции с недвижимостью        7.888225                     B  
7               образование        0.418574                     C  
8                   свадьба       18.985932                     C  
9  операции с недвижимостью        5.996593                     C

### Шаг 2.7. Категоризация целей кредита.

In [26]:
def purpose_group(purpose):
    
    if 'автомоб' in purpose:
        return 'операции с автомобилем'
    if 'недвижим' in purpose: 
        return 'операции с недвижимостью'
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образовани'in purpose:
        return 'получение образования'
    return 'другая цель' #на всякий случай, если никакая из целей не подошла

In [27]:
clients['purpose_category'] = clients['purpose'].apply(purpose_group)
clients.head(10) #появились категории для целей кредита

children  days_employed  dob_years education  education_id  \
0         1           8437       42.0    высшее             0   
1         1           4024       36.0   среднее             1   
2         0           5623       33.0   среднее             1   
3         3           4124       32.0   среднее             1   
4         0         340266       53.0   среднее             1   
5         0            926       27.0    высшее             0   
6         0           2879       43.0    высшее             0   
7         0            152       50.0   среднее             1   
8         2           6929       35.0    высшее             0   
9         0           2188       41.0   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                    purpose  years_employed total_income_category  \
0  операции с недвижимостью       23.116912                     B   
1                автомобиль       11.026860                     C   
2  операции с недвижимостью       15.406637                     C   
3               образование       11.300677                     B   
4                   свадьба      932.235814                     C   
5  операции с недвижимостью        2.537495                     B   
6  операции с недвижимостью        7.888225                     B   
7               образование        0.418574                     C   
8                   свадьба       18.985932                     C   
9  операции с недвижимостью        5.996593                     C   

           purpose_category  
0  операции с недвижимостью  
1    операции с автомобилем  
2  операции с недвижимостью  
3     получение образования  
4        проведение свадьбы  
5  операции с недвижимостью  
6  операции с недвижимостью  
7     получение образования  
8        проведение свадьбы  
9  операции с недвижимостью

### Ответы на вопросы.

##### Вопрос 1:

**1. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

In [28]:
clients['children'].value_counts()

0     14154
1      4809
2      2052
3       330
20       76
4        41
5         9
Name: children, dtype: int64

In [29]:
clients_piv1 = clients.pivot_table(index = ['children'],columns = 'debt', values = 'income_type', aggfunc = 'count')
clients_piv1['ratio'] = clients_piv1[1]/(clients_piv1[0]+clients_piv1[1])*100
clients_piv1

debt            0       1      ratio
children                            
0         13090.0  1064.0   7.517310
1          4365.0   444.0   9.232689
2          1858.0   194.0   9.454191
3           303.0    27.0   8.181818
4            37.0     4.0   9.756098
5             9.0     NaN        NaN
20           68.0     8.0  10.526316

1. начнем с того, что я не увидела расшифровки 0 и 1 для столбца debt. Допустим, что 1 нет задолженности, 0 есть.
2. у нас есть NaN для людей с 5 детьми, значит все они имеют задолженности
3. можно отметить, что с увеличением количества детей, люди реже берут кредит
4. соответвенно больше всего берут кредиты бездетные
5. по соотношению людей с задолженностями и без нет какой-то зависимости, у всех около 10%
6. при этом, если 20 детей это 2 или 0, то значения сильно не поменяются

**2. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [30]:
clients['family_status'].value_counts()

женат / замужем          12344
гражданский брак          4163
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

In [31]:
clients_piv2 = clients.pivot_table(index = ['family_status'],columns = 'debt', values = 'income_type', aggfunc = 'count')
clients_piv2['ratio'] = clients_piv2[1]/(clients_piv2[0]+clients_piv2[1])*100
clients_piv2.sort_values(by = 'ratio')

debt                       0    1     ratio
family_status                              
вдовец / вдова           896   63  6.569343
в разводе               1110   85  7.112971
женат / замужем        11413  931  7.542126
гражданский брак        3775  388  9.320202
Не женат / не замужем   2536  274  9.750890

1. после сортировки стала видна зависимость возвратов кредита
2. предполагаю, что при потере/разводе кредит становится выплачивать сложнее, поэтому процент самый низкий
3. а люди, полагающиеся только на себя (не женат/не замужем) бодрее выплачивают кредиты

**3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [32]:
#у нас нет различных групп по доходам, попробуем выделить их по процентам от минимального до максимального с шагом 25%
#clients.describe()
clients

children  days_employed  dob_years education  education_id  \
0             1           8437       42.0    высшее             0   
1             1           4024       36.0   среднее             1   
2             0           5623       33.0   среднее             1   
3             3           4124       32.0   среднее             1   
4             0         340266       53.0   среднее             1   
...         ...            ...        ...       ...           ...   
21466         1           4529       43.0   среднее             1   
21467         0         343937       67.0   среднее             1   
21468         1           2113       38.0   среднее             1   
21469         3           3112       38.0   среднее             1   
21470         2           1984       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21466  гражданский брак                 1      F   компаньон     0   
21467   женат / замужем                 0      F   пенсионер     0   
21468  гражданский брак                 1      M   сотрудник     1   
21469   женат / замужем                 0      M   сотрудник     1   
21470   женат / замужем                 0      F   сотрудник     0   

       total_income                   purpose  years_employed  \
0            253875  операции с недвижимостью       23.116912   
1            112080                автомобиль       11.026860   
2            145885  операции с недвижимостью       15.406637   
3            267628               образование       11.300677   
4            158616                   свадьба      932.235814   
...             ...                       ...             ...   
21466        224791  операции с недвижимостью       12.409087   
21467        155999                автомобиль      942.294258   
21468         89672  операции с недвижимостью        5.789991   
21469        244093                автомобиль        8.527347   
21470         82047                автомобиль        5.437007   

      total_income_category          purpose_category  
0                         B  операции с недвижимостью  
1                         C    операции с автомобилем  
2                         C  операции с недвижимостью  
3                         B     получение образования  
4                         C        проведение свадьбы  
...                     ...                       ...  
21466                     B  операции с недвижимостью  
21467                     C    операции с автомобилем  
21468                     C  операции с недвижимостью  
21469                     B    операции с автомобилем  
21470                     C    операции с автомобилем  

[21471 rows x 15 columns]

In [33]:
#def group_income(sal):
#    if 20667 <=sal <= 107654.5:
#        return 'низкий достаток'
#    if 107654.5 <sal <= 145017:
#        return 'средний достаток'
#    if 145017 <sal <= 195751.5:
#        return 'высокий достаток'
#    if 195751.5 <sal <= 2265604:
#        return 'очень высокий достаток'


In [34]:
#clients['income_status'] = clients['total_income'].apply(group_income)
#clients.head(10) #проверили новый столбец

In [35]:
clients_piv3 = clients.pivot_table(index = ['total_income_category'],columns = 'debt', values = 'total_income', aggfunc = 'count')
clients_piv3['ratio'] = clients_piv3[1]/(clients_piv3[0]+clients_piv3[1])*100
#clients_piv3.sort_values(by = 'ratio')
clients_piv3

debt                       0     1     ratio
total_income_category                       
A                         23     2  8.000000
B                       4685   356  7.062091
C                      14673  1360  8.482505
D                        329    21  6.000000
E                         20     2  9.090909

1. люди со средним достатком (группа С) чаще всех берут кредиты
2. по отношению задолженности разница не очень очевидна
3. люди с низким достатком и очень высоким достатком хуже всего возвращают кредиты, предполагаю, что первые по очевидным причинам, а вторые может как много зарабатывают, так много и тратят, поэтому им тоже сложнее возвращать кредиты.

**4. Как разные цели кредита влияют на его возврат в срок?**

In [36]:
clients['purpose_category'].value_counts()

операции с недвижимостью    10814
операции с автомобилем       4308
получение образования        4014
проведение свадьбы           2335
Name: purpose_category, dtype: int64

In [37]:
clients_piv4 = clients.pivot_table(index = ['purpose_category'],columns = 'debt', values = 'income_type', aggfunc = 'count')
clients_piv4['ratio'] = clients_piv4[1]/(clients_piv4[0]+clients_piv4[1])*100
clients_piv4.sort_values(by = 'ratio')

debt                          0    1     ratio
purpose_category                              
операции с недвижимостью  10032  782  7.231367
проведение свадьбы         2149  186  7.965739
получение образования      3644  370  9.217738
операции с автомобилем     3905  403  9.354689

1. хуже всего возвращают люди при покупке недвижимости, думаю это связано с размером кредита и длительностью выплат
2. на свадьбы хоть и тратят меньше денег, но возвращать кредит тоже не особо любят
3. обдуманные цели, такие как образование и автомобиль, показали лучший процент возврата
4. но, по сути, разница для всех целей не такая заметная

## Общий вывод:

В общем получить какую-то зависимость оказалось тяжело. Так как +- у всех схожие параметры касательно возврата кредита. Однако даже небольшие отличия могут значительно повлиять на возможности и денежный оборот банка.
Думаю, банку стоит обратить внимание на людей с семейным положением : вдовец / вдова,в разводе. Также людят с очень низким и очень высоким доходом, и при покупке недвижимости и организации свадьбы.

Более того, существенным фктором являются наличие ошибок в данных, поэтому следует:
- более тщательно следить за заполнением всех колонок
- создать правила заполнения: в каких единицах, стаж работы в годах
- создать и учитывать группировки 
- следить за дубликатами
- написать проверочный код, который уведомит о заранее невозможных данных (стаж 1000 лет, 20 детей и т.д.) и покажет ошибку 
- на основе проведенного в конце анализа рассчитать предварительный процент одобрения банком